File: apps/workbench/frontend/vite/src/api/execution.ts
import { CASES_PATH, FixtureCaseContext } from "./cases";

export const P36_EVIDENCE_FIXTURE_WORK_UNIT_TYPE = "p36_evidence_fixture_entry";
export const AGENT_FIXTURE_SHADOW_WORK_UNIT_TYPE = "agent_fixture_shadow_entry";
export const FIXTURE_NO_LEASE_FENCING_TOKEN = "fixture-no-lease";

export type CreateWorkUnitCommand = {
  work_unit_type: typeof P36_EVIDENCE_FIXTURE_WORK_UNIT_TYPE | typeof AGENT_FIXTURE_SHADOW_WORK_UNIT_TYPE;
  expected_case_version: number;
  input_head_digest: string;
  actor_ref: string;
  idempotency_key: string;
};

export type CancelWorkUnitCommand = {
  expected_work_unit_version: number;
  expected_state_version: number;
  fencing_token: typeof FIXTURE_NO_LEASE_FENCING_TOKEN;
  actor_ref: string;
  idempotency_key: string;
};

export type WorkUnitState = "pending" | "cancelled" | string;

export type WorkUnitExecutionItem = {
  work_unit_id: string;
  work_unit_version: number;
  state_version: number;
  state: WorkUnitState;
  input_head_digest: string;
};

export type WorkUnitExecutionView = {
  case_id: string;
  work_units: WorkUnitExecutionItem[];
};

export type ActivityEvent = {
  event_id: string;
  sequence: number;
  event_type: string;
  occurred_at: string;
  typed_stop?: string | null;
};

export type ActivityTraceView = {
  case_id: string;
  case_version: number;
  events: ActivityEvent[];
};

export type ResearchRunEventView = {
  event_id: string;
  sequence: number;
  event_type: string;
  occurred_at: string;
  causation_event_id?: string | null;
  details: Record<string, unknown>;
  redacted_fields: string[];
  private_chain_of_thought_included: false;
};

export type ResearchRunArtifactView = {
  artifact_version_id: string;
  artifact_type: string;
  producer_attempt_id: string;
  current_status: string;
  object_digest: string;
  input_refs: string[];
  payload: Record<string, unknown>;
  payload_exact: boolean;
  redacted_fields: string[];
};

export type S4CaseRuntimeProjection = {
  runtime_binding_digest: string;
  case_ticker: "DELL" | "MU";
  issuer_identifier: string;
  case_profile_ref: string;
  method_id: string;
  case_identity_namespace: string;
  paid_artifact_proven: false;
  human_review_completed: false;
};

export type ResearchRunProjectionItem = {
  research_run_id: string;
  research_run_version_id: string;
  work_unit_id: string;
  work_unit_type: string;
  attempt_id: string;
  execution_profile_version_ref: string;
  state: string;
  started_at: string;
  ended_at?: string | null;
  terminal_reason?: string | null;
  output_refs: string[];
  events: ResearchRunEventView[];
  artifacts: ResearchRunArtifactView[];
};

export type ResearchRunProjectionView = {
  case_id: string;
  runs: ResearchRunProjectionItem[];
  private_chain_of_thought_included: false;
};

export type S3PresentationSurfaceClaimView = {
  surface_claim_version_ref: string;
  program_cell_id: string;
  claim_text: string;
  specialist_judgment_ref: string;
  fact_statements: string[];
  evidence_refs: string[];
  numeric_refs: string[];
  graph_context_refs: string[];
  gap_codes: string[];
  what_would_change: string[];
  repair_ticket_refs: string[];
  stop_semantic: string;
  source_grade: string;
  numeric_sanity_status: string;
  official_or_estimate_flag: string;
};

export type S3PresentationWorkpaperCellView = {
  program_cell_id: string;
  cell_version_ref: string;
  surface_claim_ref: string;
  specialist_judgment_ref: string;
  decision_question: string;
  direct_answer: string;
  fact_statements: string[];
  evidence_refs: string[];
  numeric_refs: string[];
  graph_drilldown: {
    graph_edge_projection_ref: string;
    graph_authority: string;
    graph_status: string;
    source_followup_refs: string[];
    typed_gaps: string[];
    automatic_new_research: false;
  };
  gaps: string[];
  what_would_change: string[];
  repair_ticket_refs: string[];
  stop_semantic: string;
  review_status: string;
};

export type S3ThreeCellPresentationPackView = {
  presentation_pack_version_ref: string;
  presentation_pack_digest: string;
  case_id: string;
  research_run_id: string;
  execution_profile_version_ref: string;
  surface_claims: S3PresentationSurfaceClaimView[];
  workpaper: {
    artifact_ref: string;
    workpaper_version_ref: string;
    workpaper_digest: string;
    status: string;
    cell_sections: S3PresentationWorkpaperCellView[];
  };
  report: {
    artifact_ref: string;
    report_version_ref: string;
    report_digest: string;
    workpaper_artifact_ref: string;
    title: string;
    executive_answer: string;
    sections: Array<{
      section_id: string;
      program_cell_id: string;
      heading: string;
      content: string;
      surface_claim_ref: string;
      specialist_judgment_ref: string;
      evidence_refs: string[];
      numeric_refs: string[];
      boundary: string;
    }>;
    presentation_gaps: string[];
    writer_source_authority: false;
    writer_retrieval_authority: false;
    writer_external_tool_authority: false;
    model_writer_executed: false;
  };
  trace_review: {
    artifact_ref: string;
    trace_version_ref: string;
    trace_digest: string;
    workpaper_artifact_ref: string;
    report_artifact_ref: string;
    nodes: Array<{ node_ref: string; node_type: string; label: string }>;
    edges: Array<{ edge_id: string; from_ref: string; to_ref: string; relation: string }>;
    review_binding: {
      verifier_input_digest: string;
      execution_profile_version_ref: string;
      input_head_digest: string;
      analysis_as_of: string;
      artifact_refs: string[];
      bound_content_digests: string[];
      verifier_decision: string;
      human_review_status: "not_performed";
      human_decision: "not_performed";
      exact_digest_confirmation: false;
      findings: Array<{
        finding_id: string;
        layer: string;
        severity: string;
        status: string;
        affected_refs: string[];
        earliest_owner_ref: string;
        message: string;
      }>;
      review_targets: Array<{
        review_target_id: string;
        program_cell_id: string;
        surface_claim_ref: string;
        specialist_judgment_ref: string;
        artifact_refs: string[];
        source_grade: string;
        numeric_sanity_status: string;
        official_or_estimate_flag: string;
        cannot_infer: string[];
        what_would_change: string[];
        repair_ticket_refs: string[];
        stop_semantic: string;
        allowed_review_actions: string[];
        review_status: "not_performed";
      }>;
    };
  };
};

type ErrorEnvelope = {
  error?: {
    error_code?: string;
    message?: string;
    status_code?: number;
    trace_id?: string;
  };
  code?: string;
  message?: string;
  correlation_id?: string;
  detail?: unknown;
};

export class ExecutionApiError extends Error {
  readonly code: string;
  readonly statusCode: number;
  readonly traceId: string | undefined;
  readonly detail: unknown;

  constructor(input: { code: string; message: string; statusCode: number; traceId?: string; detail?: unknown }) {
    super(input.message);
    this.name = "ExecutionApiError";
    this.code = input.code;
    this.statusCode = input.statusCode;
    this.traceId = input.traceId;
    this.detail = input.detail;
  }
}

export const DEFAULT_FIXTURE_EXECUTION_CONTEXT: FixtureCaseContext = {
  tenantId: "fixture_internal",
  projectId: "workbench_internal",
  actorId: "analyst_internal",
  permissions: ["case:read", "planning:read", "execution:read", "execution:write", "activity:read"],
};

export class ExecutionApiClient {
  constructor(private readonly context: FixtureCaseContext = DEFAULT_FIXTURE_EXECUTION_CONTEXT) {}

  get actorRef(): string {
    return this.context.actorId;
  }

  async listWorkUnits(caseId: string): Promise<WorkUnitExecutionView> {
    return this.request<WorkUnitExecutionView>(`${casePath(caseId)}/work-units`);
  }

  async createWorkUnit(caseId: string, command: CreateWorkUnitCommand): Promise<WorkUnitExecutionView> {
    return this.request<WorkUnitExecutionView>(`${casePath(caseId)}/work-units`, mutationInit(command));
  }

  async cancelWorkUnit(caseId: string, workUnitId: string, command: CancelWorkUnitCommand): Promise<WorkUnitExecutionView> {
    return this.request<WorkUnitExecutionView>(
      `${casePath(caseId)}/work-units/${encodeURIComponent(workUnitId)}/cancel`,
      mutationInit(command),
    );
  }

  async getActivityTrace(caseId: string): Promise<ActivityTraceView> {
    return this.request<ActivityTraceView>(`${casePath(caseId)}/activity`);
  }

  async getResearchRunProjection(caseId: string): Promise<ResearchRunProjectionView> {
    return this.request<ResearchRunProjectionView>(`${casePath(caseId)}/execution-projection`);
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    headers.set("X-Fin-Case-Tenant", this.context.tenantId);
    headers.set("X-Fin-Case-Project", this.context.projectId);
    headers.set("X-Fin-Case-Actor", this.context.actorId);
    headers.set("X-Fin-Case-Permissions", this.context.permissions.join(","));

    let response: Response;
    try {
      response = await fetch(path, { ...init, headers });
    } catch (error) {
      throw new ExecutionApiError({
        code: "network_unavailable",
        message: "The fixture execution API is unavailable.",
        statusCode: 0,
        detail: error,
      });
    }

    const text = await response.text();
    const payload = text ? parsePayload(text) : {};
    if (!response.ok) throw toExecutionApiError(response.status, payload);
    return payload as T;
  }
}

export async function decisionSurfaceInputHeadDigest(contractVersionId: string): Promise<string> {
  const canonicalJson = JSON.stringify([contractVersionId]);
  const encoded = new TextEncoder().encode(canonicalJson);
  const digest = await crypto.subtle.digest("SHA-256", encoded);
  return Array.from(new Uint8Array(digest), (value) => value.toString(16).padStart(2, "0")).join("");
}

function casePath(caseId: string): string {
  return `${CASES_PATH}/${encodeURIComponent(caseId)}`;
}

function mutationInit(command: { idempotency_key: string }): RequestInit {
  return {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Idempotency-Key": command.idempotency_key,
    },
    body: JSON.stringify(command),
  };
}

function parsePayload(text: string): unknown {
  try {
    return JSON.parse(text) as unknown;
  } catch {
    return { detail: text };
  }
}

function toExecutionApiError(statusCode: number, payload: unknown): ExecutionApiError {
  const envelope = errorEnvelope(payload);
  const nested = errorEnvelope(envelope.detail);
  const error = envelope.error ?? nested.error;
  return new ExecutionApiError({
    code: error?.error_code ?? envelope.code ?? nested.code ?? `http_${statusCode}`,
    message: error?.message ?? envelope.message ?? nested.message ?? detailMessage(envelope.detail) ?? "Execution request failed.",
    statusCode,
    traceId: error?.trace_id ?? envelope.correlation_id ?? nested.correlation_id,
    detail: envelope.detail,
  });
}

function detailMessage(detail: unknown): string | undefined {
  if (typeof detail === "string") return detail;
  if (typeof detail !== "object" || detail === null) return undefined;
  const reason = (detail as { reason?: unknown }).reason;
  return typeof reason === "string" ? reason : undefined;
}

function errorEnvelope(value: unknown): ErrorEnvelope {
  return typeof value === "object" && value !== null ? value as ErrorEnvelope : {};
}

File: apps/workbench/backend/api/v1/execution.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from fastapi import APIRouter, BackgroundTasks, Header, HTTPException, Request, status
from pydantic import BaseModel, ConfigDict, Field

from ...application.case_service import CasePrincipal
from ...application.execution_service import (
    AGENT_FIXTURE_SHADOW_WORK_UNIT_TYPE,
    BOUNDED_AGENT_INTERNAL_WORK_UNIT_TYPE,
    CancelWorkUnitDraft,
    CreateWorkUnitDraft,
    ExecutionService,
    ExecutionServiceError,
    VT1_FENCING_TOKEN,
    VT1_WORK_UNIT_TYPE,
)
from sec_agent.workbench.api_contracts import request_trace_id


class CreateWorkUnitCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    work_unit_type: Literal[
        VT1_WORK_UNIT_TYPE,
        AGENT_FIXTURE_SHADOW_WORK_UNIT_TYPE,
        BOUNDED_AGENT_INTERNAL_WORK_UNIT_TYPE,
    ]
    expected_case_version: int = Field(ge=0)
    input_head_digest: str = Field(pattern=r"^[a-f0-9]{64}$")
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class CancelWorkUnitCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_work_unit_version: int = Field(ge=0)
    expected_state_version: int = Field(ge=0)
    fencing_token: Literal[VT1_FENCING_TOKEN]
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class WorkUnitExecutionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    work_unit_id: str
    work_unit_version: int = Field(ge=0)
    state_version: int = Field(ge=0)
    state: str
    input_head_digest: str = Field(pattern=r"^[a-f0-9]{64}$")


class WorkUnitExecutionView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    work_units: list[WorkUnitExecutionItem]


class ActivityEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_id: str
    sequence: int = Field(ge=0)
    event_type: str
    occurred_at: datetime
    typed_stop: str | None = None


class ActivityTraceView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    case_version: int = Field(ge=0)
    events: list[ActivityEvent]


class ResearchRunEventView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_id: str
    sequence: int = Field(ge=0)
    event_type: str
    occurred_at: datetime
    causation_event_id: str | None = None
    details: dict[str, Any]
    redacted_fields: list[str]
    private_chain_of_thought_included: Literal[False] = False


class ResearchRunArtifactView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    artifact_version_id: str
    artifact_type: str
    producer_attempt_id: str
    current_status: str
    object_digest: str = Field(pattern=r"^[a-f0-9]{64}$")
    input_refs: list[str]
    payload: dict[str, Any]
    payload_exact: bool
    redacted_fields: list[str]


class ResearchRunProjectionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    research_run_id: str
    research_run_version_id: str
    work_unit_id: str
    work_unit_type: str
    attempt_id: str
    execution_profile_version_ref: str
    state: str
    started_at: datetime
    ended_at: datetime | None = None
    terminal_reason: str | None = None
    output_refs: list[str]
    events: list[ResearchRunEventView]
    artifacts: list[ResearchRunArtifactView]


class ResearchRunProjectionView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    runs: list[ResearchRunProjectionItem]
    private_chain_of_thought_included: Literal[False] = False


def build_execution_router(service: ExecutionService) -> APIRouter:
    router = APIRouter(tags=["point02-execution"])

    @router.get(
        "/cases/{case_id}/work-units",
        operation_id="listWorkUnits",
        response_model=WorkUnitExecutionView,
    )
    def list_work_units(
        case_id: str,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            return service.list_work_units(
                case_id,
                _principal(tenant_id, project_id, actor_id, permissions),
            )
        except ExecutionServiceError as exc:
            _raise_service_error(exc)

    @router.post(
        "/cases/{case_id}/work-units",
        operation_id="createWorkUnit",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=WorkUnitExecutionView,
    )
    def create_work_unit(
        case_id: str,
        command: CreateWorkUnitCommand,
        request: Request,
        background_tasks: BackgroundTasks,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            principal = _principal(tenant_id, project_id, actor_id, permissions)
            trace_id = request_trace_id(request)
            view = service.create_work_unit(
                case_id,
                CreateWorkUnitDraft(**command.model_dump()),
                principal,
                trace_id=trace_id,
            )
            work_unit_id = service.pending_work_unit_id_for_type(
                case_id,
                command.work_unit_type,
                principal,
                idempotency_key=command.idempotency_key,
            )
            if service.background_dispatch_enabled and work_unit_id:
                background_tasks.add_task(
                    service.dispatch_queued_work_unit,
                    case_id,
                    work_unit_id,
                    principal,
                    trace_id=trace_id,
                )
            return view
        except ExecutionServiceError as exc:
            _raise_service_error(exc)

    @router.post(
        "/cases/{case_id}/work-units/{work_unit_id}/cancel",
        operation_id="cancelWorkUnit",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=WorkUnitExecutionView,
    )
    def cancel_work_unit(
        case_id: str,
        work_unit_id: str,
        command: CancelWorkUnitCommand,
        request: Request,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            return service.cancel_work_unit(
                case_id,
                work_unit_id,
                CancelWorkUnitDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except ExecutionServiceError as exc:
            _raise_service_error(exc)

    @router.get(
        "/cases/{case_id}/activity",
        operation_id="getActivityTrace",
        response_model=ActivityTraceView,
    )
    def get_activity(
        case_id: str,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            return service.get_activity(
                case_id,
                _principal(tenant_id, project_id, actor_id, permissions),
            )
        except ExecutionServiceError as exc:
            _raise_service_error(exc)

    @router.get(
        "/cases/{case_id}/execution-projection",
        operation_id="getResearchRunProjection",
        response_model=ResearchRunProjectionView,
    )
    def get_research_run_projection(
        case_id: str,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            return service.get_research_run_projection(
                case_id,
                _principal(tenant_id, project_id, actor_id, permissions),
            )
        except ExecutionServiceError as exc:
            _raise_service_error(exc)

    return router


def _principal(
    tenant_id: str | None,
    project_id: str | None,
    actor_id: str | None,
    permissions: str | None,
) -> CasePrincipal:
    return CasePrincipal(
        tenant_id=(tenant_id or "").strip(),
        project_id=(project_id or "").strip(),
        actor_id=(actor_id or "").strip(),
        permissions=frozenset(item.strip() for item in (permissions or "").split(",") if item.strip()),
    )


def _raise_service_error(error: ExecutionServiceError) -> None:
    raise HTTPException(status_code=error.status_code, detail=error.detail) from error
